pixeltable/catalog/catalog.py

```diff
@@ -727,11 +910,88 @@ def _load_tbl_version(self, tbl_id: UUID, effective_version: Optional[int]) -> O
             )
 
         row = conn.execute(q).one_or_none()
-        tbl_record, schema_version_record = _unpack_row(row, [schema.Table, schema.TableSchemaVersion])
+        assert row is not None, f'Table record not found: {tbl_id}:{effective_version}'
+        tbl_record, version_record, schema_version_record = _unpack_row(
+            row, [schema.Table, schema.TableVersion, schema.TableSchemaVersion]
+        )
+        assert tbl_record.id == tbl_id
         tbl_md = schema.md_from_dict(schema.TableMd, tbl_record.md)
+        version_md = schema.md_from_dict(schema.TableVersionMd, version_record.md)
         schema_version_md = schema.md_from_dict(schema.TableSchemaVersionMd, schema_version_record.md)
+
+        return schema.FullTableMd(tbl_md, version_md, schema_version_md)
+
+    def store_tbl_md(
+        self,
+        tbl_id: UUID,
+        tbl_md: Optional[schema.TableMd],
+        version_md: Optional[schema.TableVersionMd],
+        schema_version_md: Optional[schema.TableSchemaVersionMd],
+    ) -> None:
+        """
+        Stores metadata to the DB. If specified, `tbl_md` will be updated in place (only one such record can exist
+        per UUID); `version_md` and `schema_version_md` will be inserted as new records.
+
+        If inserting `version_md` or `schema_version_md` would be a primary key violation, an exception will be raised.
+        """
+        conn = Env.get().conn
+
+        if tbl_md is not None:
+            conn.execute(
+                sql.update(schema.Table.__table__)
+                .values({schema.Table.md: dataclasses.asdict(tbl_md)})
+                .where(schema.Table.id == tbl_id)
+            )
+
+        if version_md is not None:
+            conn.execute(
+                sql.insert(schema.TableVersion.__table__).values(
+                    tbl_id=tbl_id, version=version_md.version, md=dataclasses.asdict(version_md)
+                )
+            )
+
+        if schema_version_md is not None:
+            conn.execute(
+                sql.insert(schema.TableSchemaVersion.__table__).values(
+                    tbl_id=tbl_id,
+                    schema_version=schema_version_md.schema_version,
+                    md=dataclasses.asdict(schema_version_md),
+                )
+            )
+
+    def delete_tbl_md(self, tbl_id: UUID) -> None:
+        """
+        Deletes all table metadata from the store for the given table UUID.
+        """
+        conn = Env.get().conn
+        conn.execute(sql.delete(schema.TableSchemaVersion.__table__).where(schema.TableSchemaVersion.tbl_id == tbl_id))
+        conn.execute(sql.delete(schema.TableVersion.__table__).where(schema.TableVersion.tbl_id == tbl_id))
+        conn.execute(sql.delete(schema.Table.__table__).where(schema.Table.id == tbl_id))
+
+    def load_tbl_hierarchy_md(self, tbl: Table) -> list[schema.FullTableMd]:
```

Comment:
This is really specifically the replica md, so why not call it load_replica_md() ("table hierarchy" isn't used elsewhere).

Also, I think this is the correct place to do the adjustment of current_version/current_schema_version.